File: kora-java-cache-caffeine/src/main/java/ru/tinkoff/kora/example/cache/caffeine/SimpleService.java
package ru.tinkoff.kora.example.cache.caffeine;

import jakarta.annotation.Nonnull;
import java.math.BigDecimal;
import java.util.concurrent.ThreadLocalRandom;
import ru.tinkoff.kora.cache.CacheKeyMapper;
import ru.tinkoff.kora.cache.annotation.CacheInvalidate;
import ru.tinkoff.kora.cache.annotation.CachePut;
import ru.tinkoff.kora.cache.annotation.Cacheable;
import ru.tinkoff.kora.common.Component;
import ru.tinkoff.kora.common.Mapping;
import ru.tinkoff.kora.common.annotation.Root;

@Root
@Component
public class SimpleService {

    public record UserContext(String userId, String traceId) {}

    public static final class UserContextMapping implements CacheKeyMapper<String, UserContext> {

        @Nonnull
        @Override
        public String map(UserContext arg) {
            return arg.userId();
        }
    }

    @Mapping(UserContextMapping.class)
    @Cacheable(SimpleCache.class)
    public Long getMapping(UserContext context) {
        return ThreadLocalRandom.current().nextLong(0, 100_000_000L);
    }

    @Cacheable(SimpleCache.class)
    public Long get(String id) {
        return ThreadLocalRandom.current().nextLong(0, 100_000_000L);
    }

    @CachePut(value = SimpleCache.class, parameters = { "id" })
    public Long put(BigDecimal arg2, String arg3, String id) {
        return ThreadLocalRandom.current().nextLong(0, 100_000_000L);
    }

    @CacheInvalidate(SimpleCache.class)
    public void delete(String id) {}

    @CacheInvalidate(value = SimpleCache.class, invalidateAll = true)
    public void deleteAll() {}
}

File: examples/kotlin/kora-kotlin-cache-caffeine/src/main/kotlin/ru/tinkoff/kora/kotlin/example/cache/caffeine/SimpleService.kt
package ru.tinkoff.kora.kotlin.example.cache.caffeine

import jakarta.annotation.Nonnull
import ru.tinkoff.kora.cache.CacheKeyMapper
import ru.tinkoff.kora.cache.annotation.CacheInvalidate
import ru.tinkoff.kora.cache.annotation.CachePut
import ru.tinkoff.kora.cache.annotation.Cacheable
import ru.tinkoff.kora.common.Component
import ru.tinkoff.kora.common.Mapping
import ru.tinkoff.kora.common.annotation.Root
import java.math.BigDecimal
import java.util.concurrent.ThreadLocalRandom

@Root
@Component
open class SimpleService {
    data class UserContext(val userId: String, val traceId: String)

    class UserContextMapping : CacheKeyMapper<String, UserContext> {
        @Nonnull
        override fun map(arg: UserContext): String = arg.userId
    }

    @Mapping(UserContextMapping::class)
    @Cacheable(SimpleCache::class)
    open fun getMapping(context: UserContext): Long = ThreadLocalRandom.current().nextLong(0, 100_000_000L)

    @Cacheable(SimpleCache::class)
    open fun get(id: String): Long = ThreadLocalRandom.current().nextLong(0, 100_000_000L)

    @CachePut(value = SimpleCache::class, parameters = ["id"])
    open fun put(arg2: BigDecimal, arg3: String, id: String): Long =
        ThreadLocalRandom.current().nextLong(0, 100_000_000L)

    @CacheInvalidate(SimpleCache::class)
    open fun delete(id: String) = Unit

    @CacheInvalidate(value = SimpleCache::class, invalidateAll = true)
    open fun deleteAll() = Unit
}

